### PR TITLE
feat: Add state to ADW (#24)

### PR DESCRIPTION
**Pull request by paysdoc**
## Summary
Implements #24 - Add state to ADW

## Implementation Plan
I've created the implementation plan. Here's a summary:

## Summary

- **Plan created at:** `specs/issue-24-plan.md`

### What the plan covers:

- **New state directory structure**: `agents/{adwId}/orchestrator/{agent-name}/` with recursive nesting for called agents
- **Core implementation**: New `AgentStateManager` class in `adws/core/agentState.ts` with methods for:
  - Initializing state directories
  - Writing/reading `state.json` files
  - Appending to `execution.log` files
  - Writing raw JSON/JSONL output files
  - Reading parent agent state for context sharing

- **State schema**: Each agent's `state.json` contains:
  - `adwId`, `issueNumber`, `branchName`, `planFile`, `issueClass`
  - Plus execution metadata (status, timestamps, agent name, parent)

- **12 implementation steps** covering:
  1. Type definitions in `dataTypes.ts`
  2. Configuration constants in `config.ts`
  3. `AgentStateManager` class creation
  4. Utility functions in `utils.ts`
  5. `.gitignore` update with `/agents`
  6. `claudeAgent.ts` state integration
  7. `adwPlanBuild.tsx` orchestrator updates
  8. `adwPrReview.tsx` orchestrator updates
  9. `planAgent.ts` state context
  10. `buildAgent.ts` state context
  11. Unit tests
  12. Validation

- **Testing strategy**: Unit tests for AgentStateManager, integration tests for full workflow, and edge case handling

- **Validation commands**: Lint, build, test, plus manual verification of state file contents

### What was Implemented

I implemented a file-based state management system for the ADW (AI Developer Workflow) as specified in the implementation plan for GitHub Issue #24. The feature introduces:

1. **State Types** (`dataTypes.ts`):
   - `AgentIdentifier` type for consistent agent naming
   - `AgentExecutionStatus` type for tracking progress
   - `AgentExecutionState` interface for execution tracking
   - `AgentState` interface with all required fields

2. **Configuration** (`config.ts`):
   - Added `AGENTS_STATE_DIR` constant pointing to `<project root>/agents/`

3. **AgentStateManager Class** (`agentState.ts`):
   - `initializeState()` - Creates hierarchical state directories
   - `writeState()` / `readState()` - JSON state file operations with merge support
   - `appendLog()` - Timestamped execution logging with prompt capture
   - `writeRawOutput()` - JSON/JSONL raw output writing
   - `readParentState()` - Traverses up to find parent state
   - `createExecutionState()` / `completeExecution()` - Execution state helpers
   - `getStatePath()` / `stateExists()` - Path utilities

4. **Utility Functions** (`utils.ts`):
   - `ensureAgentStateDirectory()` - Directory creation
   - `getAgentStatePath()` - Path calculation without creation

5. **Agent Integration** (`claudeAgent.ts`):
   - Added optional `statePath` parameter to `runClaudeAgent()`
   - State logging for prompts, tool usage, and results
   - Raw JSONL output to state directory

6. **Orchestrator Updates**:
   - `adwPlanBuild.tsx` - Full state tracking for orchestrator, classifier, plan-agent, and build-agent
   - `adwPrReview.tsx` - Full state tracking for PR review workflows

7. **Agent Updates** (`planAgent.ts`, `buildAgent.ts`):
   - Added optional `statePath` parameters to all agent runner functions

### Files Changed/Created

**New Files:**
- `adws/core/agentState.ts` - AgentStateManager class (220 lines)
- `adws/__tests__/agentState.test.ts` - Comprehensive unit tests (30 tests)

**Modified Files:**
- `adws/core/dataTypes.ts` - Added state-related types
- `adws/core/config.ts` - Added AGENTS_STATE_DIR constant
- `adws/core/utils.ts` - Added state directory utilities
- `adws/core/index.ts` - Export new types and utilities
- `adws/agents/claudeAgent.ts` - State integration in agent runner
- `adws/agents/planAgent.ts` - Added statePath parameter
- `adws/agents/buildAgent.ts` - Added statePath parameter
- `adws/adwPlanBuild.tsx` - Full orchestrator state tracking
- `adws/adwPrReview.tsx` - Full orchestrator state tracking
- `.gitignore` - Added `/agents` entry

### Validation Results

- **Lint**: No ESLint warnings or errors
- **Build**: Compiled successfully
- **Tests**: 46 tests passed (16 existing + 30 new agent state tests)

### Directory Structure After Workflow

After running `npx tsx adws/adwPlanBuild.tsx <issue-number>`, the state directory will look like:

```
agents/
└── adw-{timestamp}-{random}/
    └── orchestrator/
        ├── state.json              # {adwId, issueNumber, branchName, planFile, issueClass}
        ├── execution.log           # Orchestrator logs
        ├── classifier/
        │   ├── state.json
        │   ├── execution.log
        │   └── output.jsonl
        ├── plan-agent/
        │   ├── state.json
        │   ├── execution.log
        │   └── output.jsonl
        └── build-agent/
            ├── state.json
            ├── execution.log
            └── output.jsonl
```

## Testing
- [ ] All validation commands from the plan pass
- [ ] No regressions introduced
- [ ] Code follows project conventions

---
Generated by ADW Plan & Build workflow

### CODE REVIEW

File: specs/issue-24-plan.md [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+
+### Step 2: Add Configuration Constants
+- Add `AGENTS_STATE_DIR` constant in `config.ts` pointing to `<project root>/agents`
+- Ensure the constant uses `process.cwd()` for proper path resolution
+
+### Step 3: Create AgentStateManager Class
+- Create `adws/core/agentState.ts` with the `AgentStateManager` class
+- Implement `initializeState(adwId, agentIdentifier, parentAgentPath?)` method
+  - Creates directory structure: `agents/{adwId}/{agentIdentifier}/`
+  - For nested agents: `agents/{adwId}/{parentAgent}/{agentIdentifier}/`
```
Comment (paysdoc) [inline specs/issue-24-plan.md:81]: issue-24-plan.md reuired sub-agents to have structure `agents/{adwId}/{parentAgent}/{agentIdentifier}/`. However, the structure `agents/{parentAdwId}/{adwId}/{agentIdentifier}/` is clearer

**Comment by paysdoc** [comment]
## :eyes: Addressing PR Review Comments

Starting automated review of **1** review comment(s).

**ADW ID:** `adw-1770036875923-rpsx0m`

_Analyzing review feedback..._

**Comment by paysdoc** [comment]
## :eyes: Addressing PR Review Comments

Starting automated review of **1** review comment(s).

**ADW ID:** `adw-1770036876540-7o4u1g`

_Analyzing review feedback..._

**Comment by paysdoc** [comment]
## :pencil: Planning PR Review Changes

Creating a revision plan based on review feedback...

**ADW ID:** `adw-1770036875923-rpsx0m`

**Comment by paysdoc** [comment]
## :pencil: Planning PR Review Changes

Creating a revision plan based on review feedback...

**ADW ID:** `adw-1770036876540-7o4u1g`

**Comment by paysdoc** [comment]
## :white_check_mark: Revision Plan Created

A revision plan has been created to address the review comments.

**ADW ID:** `adw-1770036875923-rpsx0m`

<details>
<summary>Revision Plan Summary</summary>

Now I have a complete understanding of the current implementation and the requested change. Let me create the revision plan.

---

# Revision Plan: Update State Directory Structure

## Summary

This revision addresses the PR review comment requesting a clearer directory structure for sub-agents. The change updates the structure from `agents/{adwId}/{parentAgent}/{agentIdentifier}/` to `agents/{parentAdwId}/{adwId}/{agentIdentifier}/`.

## Review Comment Analysis

**Comment:** The reviewer suggests that the structure `agents/{parentAdwId}/{adwId}/{agentIdentifier}/` is clearer than the currently implemented `agents/{adwId}/{parentAgent}/{agentIdentifier}/`.

**Interpretation:** Each agent invocation should have its own ADW ID, and the directory hierarchy should be based on ADW ID relationships rather than agent names. This provides:
- Better traceability of agent call chains through unique IDs
- Clearer parent-child relationships in the file system
- Consistent use of ADW IDs at each nesting level

## Directory Structure Comparison

### Current Structure
```
agents/
└── adw-1234567890-abc123/           # Single shared ADW ID
    └── orchestrator/                 # Parent agent name
        ├── state.json
        ├── classifier/               # Child agent name
        │   └── state.json
        ├── plan-agent/
        │   └── state.json
        └── build-agent/
            └── state.json
```

### Proposed Structure
```
agents/
└── adw-1234567890-abc123/           # Orchestrator's ADW ID (root)
    ├── orchestrator/                 # Orchestrator agent directory
    │   ├── state.json
    │   └── execution.log
    ├── adw-classifier-def456/        # Classifier's own ADW ID
    │   └── classifier/               # Classifier agent directory
    │       ├── state.json
    │       └── execution.log
    ├── adw-plan-agent-ghi789/        # Plan agent's own ADW ID
    │   └── plan-agent/
    └── adw-build-agent-jkl012/       # Build agent's own ADW ID
        └── build-agent...

</details>

**Comment by paysdoc** [comment]
## :hammer_and_wrench: Implementing Review Changes

Running Build Agent to implement the revision plan...

**ADW ID:** `adw-1770036875923-rpsx0m`

**Comment by paysdoc** [comment]
## :white_check_mark: Revision Plan Created

A revision plan has been created to address the review comments.

**ADW ID:** `adw-1770036876540-7o4u1g`

<details>
<summary>Revision Plan Summary</summary>

{"type":"assistant","message":{"model":"claude-opus-4-5-20251101","id":"msg_011GUhfP3imEvd9TZNtTcrDZ","type":"message","role":"assistant","content":[{"type":"thinking","thinking":"The user wants me to create a revision plan to address a PR review comment. Let me analyze the comment:\n\n**Review Comment:**\n- File: `specs/issue-24-plan.md` (line 81)\n- Author: paysdoc\n- Comment: \"issue-24-plan.md required sub-agents to have structure `agents/{adwId}/{parentAgent}/{agentIdentifier}/`. However, the structure `agents/{parentAdwId}/{adwId}/{agentIdentifier}/` is clearer\"\n\nSo the reviewer is suggesting a different directory structure for sub-agents. The current plan has:\n```\nagents/{adwId}/{parentAgent}/{agentIdentifier}/\n```\n\nBut the reviewer prefers:\n```\nagents/{parentAdwId}/{adwId}/{agentIdentifier}/\n```\n\nWait, let me re-read this. The comment says the current plan has `agents/{adwId}/{parentAgent}/{agentIdentifier}/` but the reviewer thinks `agents/{parentAdwId}/{adwId}/{agentIdentifier}/` is clearer.\n\nActually, looking at the comment again, it seems like:\n- Current: `agents/{adwId}/{parentAgent}/{agentIdentifier}/`\n- Proposed: `agents/{parentAdwId}/{adwId}/{agentIdentifier}/`\n\nBut wait, this doesn't quite make sense. Let me look at the original plan again.\n\nFrom the original plan:\n- `agents/{adwId}/orchestrator/` for orchestrator state\n- `agents/{adwId}/orchestrator/classifier/` for classifier agent\n- `agents/{adwId}/orchestrator/plan-agent/` for plan agent\n- etc.\n\nSo the structure is: `agents/{adwId}/{orchestrator}/{agentIdentifier}/`\n\nThe reviewer's suggestion of `agents/{parentAdwId}/{adwId}/{agentIdentifier}/` seems to suggest that:\n- Each agent (even sub-agents) should have their own adwId\n\n\n- The parent's adwId would be at the top level\n- Then the current agent's adwId\n- Then the agent identifier\n\nThis makes more sense—it treats each agent as a first-class entity with its own identifier, rather than nesting them under a pa...

</details>

**Comment by paysdoc** [comment]
## :hammer_and_wrench: Implementing Review Changes

Running Build Agent to implement the revision plan...

**ADW ID:** `adw-1770036876540-7o4u1g`
